### DISCUSSION TITLE
Automatic status code and type validation from decorator fields

### DISCUSSION THREAD

**Discussion by Giurzo** [fastapi/fastapi#9884] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
class ItemStatus(BaseModel):
    description: str

class DifferentResponse(BaseModel):
    example_field: int


@app.get(
    "/item/{id}/status",
    response_model=ItemStatus,
    responses={"400": {"model": DifferentResponse}},
    status_code=200,
)
def get_status(id: str) -> Any:
    # random condition
    if len(id) > 10:
        return ItemStatus(description='item description')
    return JSONResponse(status_code=400, content={'example_field': 111})
```

### Description

In the documentation is written that if i want [additional responses](https://fastapi.tiangolo.com/advanced/additional-responses/) i have to specify them in the decorator's "responses" field and return them as a JSONResponse. 
My question is: there is a way to simply return the object as it is like
```python
return DifferentResponse(example_field=111)
```
and set the status code (400 in this case) based on the responses field?
This will also allow you to set the correct return type in the function, such as
```python
def get_status(id: str) -> Union[ItemStatus, DifferentResponse]:
``` 
instead of  
```python
def get_status(id: str) -> Any:
``` 

### Operating System

Windows

### FastAPI Version

0.100.0

### Python Version

3.11.0

**Comment by YuriiMotov** [comment]
Will just add the test case for this:

```python
from fastapi import FastAPI, Response
from fastapi.testclient import TestClient
from pydantic import BaseModel


app = FastAPI()


class ItemStatus(BaseModel):
    description: str


class DifferentResponse(BaseModel):
    example_field: int


@app.get(
    "/item/{id}/status",
    response_model=ItemStatus,
    responses={"400": {"model": DifferentResponse}},
    status_code=200,
)
def get_status(id: str, response: Response) -> ItemStatus | DifferentResponse:
    if id == 1:  # Some condition
        return ItemStatus(description="item description")

    response.status_code = 400
    return DifferentResponse(example_field=123)

# ==========================================================================
# Test

def test_400():
    client = TestClient(app, raise_server_exceptions=False)
    resp = client.get("/item/2/status")
    assert resp.status_code == 400  # AssertionError (assert 500 == 400)
    assert resp.json() == {"example_field": 123}

```
